drivers/cc1200/cc1200-rxtx.c

```diff
@@ -0,0 +1,430 @@
+        cc1200_strobe(dev, CC1200_STX);
+        //xtimer_usleep(200);
+    }
+
+    if (to_send < left) {
+        /* set GDO2 to 0x2 -> will deassert at TX FIFO below threshold */
+        cc1200_write_reg(dev, CC1200_IOCFG2, CC1200_CFG_TXFIFO_THR);
+        gpio_irq_enable(dev->params.int_pin);
+    }
+    else {
+        /* set GDO2 to 0x6 -> pin 2 will deassert at end of TX packet */
+        cc1200_write_reg(dev, CC1200_IOCFG2, CC1200_CFG_PKT_SYNC_RXTX);
+        gpio_irq_enable(dev->params.int_pin);
+    }
+}
+
+void cc1200_isr_handler(cc1200_t *dev, void (*callback)(void *), void *arg)
+{
+    uint8_t rxbytes;
+
+    switch (dev->radio_state) {
+        case RADIO_RX:
+            rxbytes = cc1200_read_reg(dev, CC1200_NUM_RXBYTES);
+            //DEBUG("radio rx\n");
+            if (gpio_read(dev->params.int_pin) | (rxbytes > 0)) {
+                //DEBUG("cc1200_isr_handler((): starting RX\n");
+                dev->isr_cb = callback;
+                dev->isr_cb_arg = arg;
+                _rx_start(dev);
+
+                /*
+                 * Don't know, why this waiting is necessary, but the driver
+                 * blocks without this. Alternatively, the delay can also be
+                 * located in  _netdev_cc1200_isr() or _isr(), but this here
+                 * is more specific. Also, the delay can be placed before or
+                 * after, but not in _rx_start()...
+                 */
+                xtimer_spin(xtimer_ticks_from_usec(800));
+            }
+            else {
+                DEBUG("cc1200_isr_handler((): isr handled too slow or falling edge trigger\n");
+                _rx_abort(dev);
+            }
+            break;
+        case RADIO_RX_BUSY:
+            //DEBUG("radio rx busy\n");
+            _rx_continue(dev, callback, arg);
+            break;
+        case RADIO_TX_BUSY:
+            //DEBUG("radio tx busy\n");
+            if (!gpio_read(dev->params.int_pin)) {
+                _tx_continue(dev);
+            }
+            else {
+                //DEBUG("cc1200_isr_handler() RADIO_TX_BUSY + GDO2\n");
+            }
+            break;
+        default:
+            DEBUG("%s:%s:%u: unhandled mode\n", RIOT_FILE_RELATIVE,
+                  __func__, __LINE__);
+            break;
+    }
+}
+
+int cc1200_send(cc1200_t *dev, const struct iovec *vector, unsigned count)
+{
+    DEBUG("%s:%u\n", __func__, __LINE__);
+
+    switch (dev->radio_state) {
+        case RADIO_RX_BUSY:
+            DEBUG("%s: rx busy...\n", __func__);
+            return -EAGAIN;
+        case RADIO_TX_BUSY:
+            DEBUG("%s: tx busy...\n", __func__);
+            //return -EAGAIN;
+            {
+                int tx_bytes = cc1200_get_reg_robust(dev, CC1200_NUM_TXBYTES);
+
+                while (tx_bytes > 0) {
+                    DEBUG("%s: waiting for transmission to be finished, %d bytes left\n",
+                          __func__, tx_bytes);
+                    thread_yield();
```

Comment:
I assume (withou checking) you yield here cause you expect a TX interrupt or something? Why do you need to spin below?

Reply:
That is correct. The spin was also included to give the transmission time to end. As we are using the sub-1GHZ band for transmission, it would occur that the processor would be faster than the transmission, hence the wait.